sklearn/preprocessing/data.py

```diff
@@ -3147,3 +3150,216 @@ def inverse_transform(self, X):
+                                " 'auto', int or array of ints, got %r"
+                                % self.n_values)
+            if n_values.ndim < 1 or n_values.shape[0] != X.shape[1]:
+                raise ValueError("Shape mismatch: if n_values is an array,"
+                                 " it has to be of shape (n_features,).")
+
+        self.n_values_ = n_values
+        n_values = np.hstack([[0], n_values - 1])
+        indices = np.cumsum(n_values)
+        self.feature_indices_ = indices
+
+        mask = (X >= self.n_values_).ravel()
+        if np.any(mask):
+            if self.handle_greater == 'error':
+                raise ValueError("handle_greater='error' but found %d feature"
+                                 " values which exceeds n_values."
+                                 % np.count_nonzero(mask))
+
+        return X
+
+    def _transform(self, X):
+        """Assumes X contains only ordinal features."""
+        X = check_array(X, dtype=np.int)
+        if np.any(X < 0):
+            raise ValueError("X needs to contain only non-negative integers.")
+        n_samples, n_features = X.shape
+
+        indices = self.feature_indices_
+        if n_features != indices.shape[0] - 1:
+            raise ValueError("X has different shape than during fitting."
+                             " Expected %d, got %d."
+                             % (indices.shape[0] - 1, n_features))
+
+        # We clip those ordinal features of X that are greater than n_values_
+        # using mask if self.handle_greater is "clip".
+        # This means, the row_indices and col_indices corresponding to the
+        # greater ordinal feature are all filled with ones.
+        mask = (X >= self.n_values_).ravel()
+        if np.any(mask):
+            if self.handle_greater == 'warn':
+                warnings.warn("Found %d feature values which exceeds "
+                              "n_values during transform, clipping them."
+                              % np.count_nonzero(mask))
+            elif self.handle_greater == 'error':
+                raise ValueError("handle_greater='error' but found %d feature"
+                                 " values which exceeds n_values during "
+                                 "transform." % np.count_nonzero(mask))
+
+        X_ceil = np.where(mask.reshape(X.shape), self.n_values_ - 1, X)
+        column_start = np.tile(indices[:-1], n_samples)
+        column_end = (indices[:-1] + X_ceil).ravel()
+        column_indices = np.hstack([np.arange(s, e) for s, e
+                                   in zip(column_start, column_end)])
+        row_indices = np.repeat(np.arange(n_samples, dtype=np.int32),
+                                X_ceil.sum(axis=1))
+        data = np.ones(X_ceil.ravel().sum())
+        out = sparse.coo_matrix((data, (row_indices, column_indices)),
+                                shape=(n_samples, indices[-1]),
+                                dtype=self.dtype).tocsr()
```

Comment:
Maybe not too important, but it is also possible to directly create a csr matrix instead of first constructing coo and then converting (for CategoricalEncoder I edited the construction of the indices to do this: https://github.com/scikit-learn/scikit-learn/pull/9151/commits/85cf315f0847d09f0f9c21180a1a9cc3cdcc3570)